### ISSUE TITLE
An open buffer for a nonexistent `foo.rs` shadows `foo/mod.rs`, emptying the module

### ISSUE THREAD

**Issue by stephanemagnenat** [rust-lang/rust-analyzer#23366]
## Summary

If the client opens a buffer (`textDocument/didOpen`) for `src/ast.rs` while that file does **not** exist on disk, rust-analyzer resolves `mod ast;` to the in-memory buffer instead of the real `src/ast/mod.rs`. If the buffer is empty, the whole module disappears for analysis, which causes many false diagnostics elsewhere. For example, glob-imported enum variants become bindings:
"Variable `Error` should have snake_case name". Types from the module become `{unknown}`, giving false `E0308` mismatches.

rustc never sees the buffer, so the crate builds fine. With both files on disk, rustc would reject it with E0761 ("file for module found at both"). rust-analyzer does not report anything like that for the buffer; it silently picks it.

The diagnostics persist as long as the buffer is open, and come back after every server restart if the editor restores the tab.

## Reproduction

Repository: https://github.com/stephanemagnenat/ra-ghost-module

```sh
git clone https://github.com/stephanemagnenat/ra-ghost-module
cd ra-ghost-module
python3 probe.py        # or RUST_ANALYZER=/path/to/rust-analyzer python3 probe.py
```

`src/ast/mod.rs`:

```rust
pub enum ExprKind {
    Literal(i64),
    Error,
}

pub struct Module {
    pub items: Vec<ExprKind>,
}
```

`src/lib.rs`:

```rust
mod ast;

pub fn is_error(module: &ast::Module, index: usize) -> bool {
    use ast::ExprKind::*;
    let item = &module.items[index];
    match item {
        Literal(_) => false,
        Error => true,
    }
}
```

`cargo check` passes.

LSP steps performed by `probe.py`:

1. `initialize` / `initialized` with the crate as root.
2. `textDocument/didOpen` for `src/lib.rs` (its disk contents).
3. `textDocument/didOpen` for `src/ast.rs` with `"text": ""`. The file does not exist on disk.
4. Wait for quiescence, then `textDocument/diagnostic` for `src/lib.rs`.

**Expected:** no diagnostics (as without step 3). A dedicated diagnostic for the ambiguous module file would also be reasonable.

**Actual:**

```
8:9 [non_snake_case] Variable `Error` should have snake_case name, e.g. `error`
```

Without step 3, the same server returns 0 diagnostics for `src/lib.rs`.

## How editors trigger it

Zed (1.19.2) restores tabs for files that were deleted or moved, such as `src/ast.rs` after a refactor to `src/ast/mod.rs`. It sends `didOpen` with empty text for each of them. In a real project Zed reopened 39 such tabs on startup, four of which collided with `mod.rs` modules. There, opening just one of them (an empty `src/ast.rs`) reproduced all the false diagnostics, and opening all real files without the ghost tabs gave none.

## Possible directions

- When both `foo.rs` and `foo/mod.rs` are candidates, prefer the file that exists on disk over a client-only buffer, or report E0761-like ambiguity instead of silently choosing.
- Alternatively, do not let buffers for paths that do not exist on disk participate in module resolution.

## Versions

- rust-analyzer 0.3.3049-standalone (release 2026-09-14, commit 682a84e95b5a)
- rustc 1.98.0 (88d9e12ae 2026-08-18)
- Zed 1.19.2 (df181c6f58d0)
- Linux x86_64

Related: #19709 (spurious "`None` should have snake_case name" in Zed), possibly #19671 (file mapping not updated when moving between `foo.rs` and `foo/mod.rs`).

**Comment by stephanemagnenat** [comment]
This happens often when using the Zed editor, see zed-industries/zed#64231

**Comment by ChayimFriedman2** [comment]
Did you use AI for this? Please read [our AI policy](https://github.com/rust-lang/rust-analyzer/blob/master/AI_POLICY.md). In particular:

> **AI should not be used to generate comments when communicating with maintainers.** We expect comments on our projects to be written by humans. We may hide any comments that we believe are AI generated.
>
> If you are opening an issue, we expect you to describe the problem in your own words.

**Comment by ChayimFriedman2** [comment]
This is either a client bug or a user error, but not a rust-analyzer bug. We correctly treat open files as existing.

**Comment by stephanemagnenat** [comment]
> Did you use AI for this? Please read [our AI policy](https://github.com/rust-lang/rust-analyzer/blob/master/AI_POLICY.md). In particular:

Yes I used AI. I did not read the policy indeed, sorry for that.

> This is either a client bug or a user error, but not a rust-analyzer bug. We correctly treat open files as existing.

I see. It affects several people though, see #19709? I am not the first one experiencing this bug.

So please, this bug has been very annoying for me and probably for others as well, I spent significant time trying to solve it, and finally I managed to find a reproducible test case, with the help of the AI, but it still took me a lot of my time.

Now you are closing this bug roughly. This is not helpful. AI have pros and cons in software development, but they are very useful to investigate subtle bugs in the specific settings of the users where manual search would not be realistically possible. I do not feel that just rejecting a possible reproducible test case for a long-standing bug (#19709), even if the report was done incorrectly, is not helping the Rust community, nor the open source world, go forward.

**Comment by ChayimFriedman2** [comment]
We do not totally forbid using AI to solve and/or analyze bugs. We do forbid using it to write issue descriptions. Please read the policy.

> Then how do you explain https://github.com/rust-lang/rust-analyzer/issues/19709? I am not the first one experiencing this bug.

That issue is caused by many different reasons for different people. Perhaps yours is one of them, but that doesn't mean this case is a rust-analyzer bug.

> So please, this bug has been very annoying for me and probably for others as well,

I don't fully understand how exactly Zed behaves and why. You may try to file a bug against them. What I do know is that when the client sends a `textDocument/didOpen` request for some path, we should consider it to exist, so that's what we do. Therefore rust-analyzer behaves as it should here, and there is no bug.

**Comment by stephanemagnenat** [comment]
> Please read the policy

I did after reading your comment, and will follow it in the future. At the same time, it is unrealistic to expect everyone reads all documents before doing anything. I was not aware of the policy when letting the LLM write this bug report.

> That issue is caused by many different reasons for different people. Perhaps yours is one of them, but that doesn't mean this case is a rust-analyzer bug.

That is why I used the LLM to build a reproducible test case based on my usage logging data. I first let the LLM to build a small proxy between Zed and rust-analyzer to capture what was going on, and based on that build a minimal reproducible test case.

The result is a small tool that can reproduce the problem without Zed: https://github.com/stephanemagnenat/ra-ghost-module

But I also see the point of rust-analyzer not wanting to re-check if files exist and leaving that up to the client to only report what exists.

> You may try to file a bug against them.

I created a bug towards Zed linking this issue: https://github.com/zed-industries/zed/issues/64231
It provides a repeatable reproducible recipe using Zed.
